### ISSUE TITLE
Clock Formatting For %0I

### ISSUE THREAD

**Issue by bisby** [Alexays/Waybar#1469]
Expected Behavior:
The clock module links to https://howardhinnant.github.io/date/date.html#to_stream_formatting - This says

> %I The hour (12-hour clock) as a decimal number.  If the result is a single digit, it is prefixed with 0.  The modified command %OI produces the locale's alternative representation.

Previously `%0I` returned `8` for 8pm, whereas `%I` returns `08`. So I would expect to have this display `8:45pm`

Actual Behavior:
The clock module returns nothing. Empty space on the bar. I can still see the module underline. When I click on the module it shows the full time. When I click on the module again it doesn't clear the full time, but full time stops updating.

Relevant Config:
```
    "clock": {
        "format": "{:%0I:%M %p}",
```

Relevant Log lines:
```
[2022-03-08 20:38:07.854] [error] clock: invalid format
```

This works when the `%0I` is changed to `%I` , but then the formatting isn't quite what I want. I only consider this an issue though because the linked documentation for date formatting says this should be valid.

**Comment by LukashonakV** [comment]
At least the issue comes from the date utility and the clock module displays info which is provided by the date library.
1. According the documentation: 

> The hour (12-hour clock) as a decimal number. If the result is a single digit, it is prefixed with 0. The modified command %OI produces the locale's alternative representation.
I don't see any mentions about the library should cut the leading zeros. I see it depends on you LC_TIME environment variable.

2. Try to do the same in terminal:
a. `date +%0I`
b. `date +%OI`
c. `date +%I`
Result on my machine:
![ps_2022-03-26-14_09_42](https://user-images.githubusercontent.com/23121044/160236870-ee71a63f-5388-47f0-8d9c-ae0b7dd86327.png)

So at least I would check you local configuration to make sure in terminal local date utility shows the output you've expecting.

**Comment by bisby** [comment]
Interesting. It had previously worked so in my brain I registered the locale's alternative as meaning "no leading 0"

It seems the easiest way with `date` to remove the 0 is `date +%-I` but that isnt valid in the library waybar uses.

I also don't have any locales set up that LC_TIME will produce a non leading zero.

Seems to be an issue for me and not waybar. Though I would love to have the option to force no-prefix

**Comment by jayache80** [comment]
Is it agreed that this is indeed a bug? And, is there a known workaround?

My waybar-clock config broke recently because I use `%-I` that will yield single-digit hours without a leading zero. However I now must use `%I` which yields an undesirable leading zero, else I have an invalid waybar-clock config.

If the intent of the date library that waybar-clock uses is to mimic the functionality of the `date` command, I'd say it's not currently doing that, because:
```
date +%-I
```
works just fine to provide a single digit hour without leading zero (and `%-I` used to work just fine in waybar-clock until recently).

**Comment by LukashonakV** [comment]
Not sure I understand correct.
Please take a look at my picture:
1. Please see ate the top waybar which I ran for an example with the changed time.
2. Please take a look a the config is opened on the left window.
For me it looks like working correct
![ps_2022-04-11-12_19_51](https://user-images.githubusercontent.com/23121044/162708030-774467f0-badf-4696-a9b2-f9db20196ec7.png)

**Comment by bisby** [comment]
![image](https://user-images.githubusercontent.com/619700/162842963-b235ba9d-31a4-46b2-8c39-9b11c454bb23.png)


For me i get in my logs `[error] clock: invalid format` and the clock shows up as an empty space. This works fine when the only change is `%I` but then i have the leading zeroes.

When running with `waybar -l trace` I get:
> [2022-04-11 17:22:44.982] [warning] As using a timezone, some format args may be missing as the date library haven't got a release since 2018.

But nothing else clock related.

**Comment by jayache80** [comment]
> 1. You are using waybar ver. 0.9.12 ?

I tried both 0.9.12 and waybar-git-r1906.503fe9a-1 (arch linux AUR package), and have the same result.

> 2. Can you check waybar config, maybe by the mistake you filled clock block, especially "format" parameter twice or more times ?

Having this `clock` block in my waybar config:
```
    "clock": {
        // "timezone": "America/New_York",
        "tooltip-format": "<big>{:%Y %B}</big>\n<tt><small>{calendar}</small></tt>",
        "format": "{:%I}"
    },
```
will work to show a leading 0 hour only.

Having this in the `clock` block of my waybar config:
```
    "clock": {
        // "timezone": "America/New_York",
        "tooltip-format": "<big>{:%Y %B}</big>\n<tt><small>{calendar}</small></tt>",
        "format": "{:%-I}"
```
will break, and then I get nothing in the clock area.

As @bisby has alluded, is this perhaps caused by a timezone locale configuration, a system date library thing, or some other system config thing?

Using the `clock` block's `timezone` field yielded even stranger results.

This `clock` block:
```
    "clock": {
        "timezone": "America/New_York",
        "tooltip-format": "<big>{:%Y %B}</big>\n<tt><small>{calendar}</small></tt>",
        "format": "{:%I}"
    },
```
yields a leading 0 hour (in NY time, a different timezone than mine. Neat.)

This `clock` block:
```
    "clock": {
        "timezone": "America/New_York",
        "tooltip-format": "<big>{:%Y %B}</big>\n<tt><small>{calendar}</small></tt>",
        "format": "{:%-I}"
    },
```
strangely yields a literal `%-I` in the clock area.

**Comment by LukashonakV** [comment]
> As @bisby has alluded, is this perhaps caused by a timezone locale configuration, a system date library thing, or some other system config thing?

I'm using GNU coreutils with the fmt package version - 8.32-r1. Can you check is the same version is installed on your machine or you're using another one or even another fmt package. Actually according the code this fmt package handles text formatting.

**Comment by bisby** [comment]
> fmt (GNU coreutils) 9.0

**Comment by jayache80** [comment]
`fmt 8.1.1-2`

**Comment by jayache80** [comment]
I made a request to fmtlib:
https://github.com/fmtlib/fmt/issues/2959

Either some `strftime` features are no longer fully implemented in fmt (as `%-I` used to work on Waybar and now it doesn't, but Waybar has always used fmt from what I gather) or the usage is now different and there's now user error.

**Comment by hajosattila** [comment]
OFF
[LukashonakV](https://github.com/LukashonakV)
please share your waybar css and config file! (very nice)

**Comment by LukashonakV** [comment]
Hi @hajosattila , config is [Dotfiles](https://gitlab.com/Viktar/dotfiles/-/blob/main/.config/waybar/config)

**Comment by LukashonakV** [comment]
Hi @bisby is this still actual ?

**Comment by bisby** [comment]
I've been back on i3wm so not using waybar recently. but booting it up to check...

```
[2023-01-27 15:28:12.955] [error] clock: invalid format
```

for

```
"clock": {$                  
    "format": "{:%-I:%M %p}",
```

So yes. clock formatting with no leading zeroes is seemingly still not possible in my locale.
And I'm now on `fmt (GNU coreutils) 9.1`

**Comment by iambumblehead** [comment]
I want my waybar to look like gnome's date time area --no leading '0' on the month number

<img width="300px" src="https://user-images.githubusercontent.com/2058705/220201069-a67e33a8-d787-42f0-8b87-3abbb12b5e79.png" />

**Comment by jayache80** [comment]
Upstream fmt has brought in a commit to master (but not in any releases yet) that should resolve this issue:

https://github.com/fmtlib/fmt/commit/7718eeeacc98c2dfb6fe8714fb38ef31f2921ab9

I had trouble building waybar with latest fmt (problems with spdlog) so I couldn't test yet.

Also, these are related and possibly the same underlying issue:
https://github.com/Alexays/Waybar/issues/1696
https://github.com/Alexays/Waybar/issues/1447

**Comment by LukashonakV** [comment]
Hi @jayache80 , @bisby according upstream this feature was implemented by the [PR#3271](https://github.com/fmtlib/fmt/pull/3271) . I've checked the source code and looks like once fmt will be official release Waybar can use this feature. 
For example see: [test/chrono-test.cc](https://github.com/fmtlib/fmt/pull/3271/files#diff-1ee69b67ff4425f4dfb5595c0eee868c39579a3ad8cdc48d51817c4ffc4b5179R956)

**Comment by kevoh1516** [comment]
Stumbled onto this issue while googling how to remove leading zero. Is this still an issue a year later?

**Comment by hhhorb** [comment]
Still an issue in today's release...
```
❯ waybar -v
Waybar v0.10.1 (branch 'master')
```
Module formatting...
```
    "clock": {
         "format": "{:%OI:%M}"
         }
```

Expected output (example)...
```
9:45
```
Actual output...
```
09:45
```

**Comment by hhhorb** [comment]
Still an issue on v0.10.2.

**Comment by jayache80** [comment]
I don't believe fmt uses `O` alone to strip the leading zeroes, but rather `-`. So to get
```
9:45
```
you'd want to use:
```
{:%-I:%M}
```

Anywho, this will be fixed in waybar once
1. fmt cuts a new release, as the latest fmt release doesn't contain [this commit](https://github.com/fmtlib/fmt/commit/470c4e6ca8d648a70f0071fa7c81b15fbc4556a4) which gets `-` working correctly to strip leading zeroes where the user requests (latest release has [this bug](https://github.com/fmtlib/fmt/issues/3811) which causes all leading zeroes to get stripped, not just the requested ones)
2. meson wrapdb gets that new release
3. waybar updates the dependency

**Comment by trevin-j** [comment]
This issue has been very frustrating for me because it seems so simple. For a temporary workaround until the fmt side of things is worked out, you can output the `date` command in a custom module e.g.:

```
    "custom/clock": {
        "exec": "date +\"%a %b %e %-I:%M %p\"",
        "interval": 1
    },
```

In the jsonc file and add `"custom/clock"` to one of the modules list.

**Comment by hhhorb** [comment]
@trevin-j, now why the hell does that work? Isn't it still calling fmt/libfmt? Anyway, THANK YOU for the workaround!

**Comment by craethke** [comment]
The fix was released in fmt version 11.0.0: https://github.com/fmtlib/fmt/releases/tag/11.0.0

Meson WrapDB has versions 11.0.1-1 and 11.0.2-1 available: https://mesonbuild.com/Wrapdb-projects.html

Any other blockers to pulling in the updated dependency?

**Comment by ShouxTech** [comment]
Still no simple way to remove the leading zero from the clock!
Why is this still an issue?

**Comment by benjohnson77** [comment]
Up voting the clean fix ... really cool to see the communities persistence.

**Comment by EricWardner** [comment]
@craethke looks like it did get pulled in? Could be reading this wrong.

https://github.com/Alexays/Waybar/blob/0776e694df56c2c849b682369148210d81324e93/subprojects/fmt.wrap#L7

**Comment by rehanzo** [comment]
The fact that running the date command every second could have the clock off by a second, as well as the fact that we run it every second when we know we don't need it (if your clock just shows hour and minute) annoyed me with the solution above. So I went ahead and overengineered a solution:

```jsonc
// default-modules.jsonc
  "custom/clock": {
    "exec": "~/Scripts/waybar-time.sh 2>/dev/null",
    "format": "<span rise='0'>{}</span>",
    "tooltip": false,
    "restart-interval": 1
  }
```

```sh
# waybar-time.sh

#!/bin/bash

print_time() {
    local_date=$(date "+%a %-d")
    local_time=$(date "+%-I:%M")
    combined="${local_date}  ·   ${local_time}"
    echo "$combined"
}

print_time

trap "true" SIGRTMIN+1

counter=0

while true; do
    # reset timing every 10 loops or after signal interrupt
    if [ $counter -eq 0 ]; then
        now=$(date +%S.%N)
        sleep_time=$(awk -v now=$now 'BEGIN { print 60 - now }')
    else
        sleep_time=60
    fi

    sleep $sleep_time &
    wait_result=$?

    # if signal to reset, reset counter
    if [ $wait_result -ne 0 ]; then
        counter=0 
    else
        counter=$(( (counter + 1) % 10 ))
    fi

    print_time
done
```

Also set this to run everytime my system resumes from suspension (since sleep doesn't count while its suspended):
```
{ sleep 1; pkill -SIGRTMIN+1 waybar-time.sh; } &
```

Threw this in `/etc/zzz.d/resume/resume.sh` since I'm on voidlinux. This solution will only run the date command when needed.

Not sure if it was ever just an issue on my end, but I do recall the default clock modules sometimes being inaccurate when I put my system to sleep. This doesn't have such an issue. One can change the formatting in the script to whatever.

**Comment by Samq64** [comment]
Is there a status update on this? From my understanding Waybar has been built with a version of `fmt` that supports `%-I` for a while now but it doesn't work for me on v0.14.0.

**Comment by ugoa** [comment]
Now I feel the Arch/Hyprland vibes...

**Comment by jayache80** [comment]
> From my understanding Waybar has been built with a version of fmt that supports %-I for a while now but it doesn't work for me on v0.14.0.

Agree. I still have an issue with this:

```
    "clock": {
        "format": "{:%A  %b %e, %Y  %-I:%M %p}"
    },
```

which gives me this error message when starting Waybar (and an empty clock):

    [error] clock: chrono format error: invalid  specifier in chrono-specs

The call to `vformat` here is the first place in the clock.cpp (the clock module), around line 165, where I hit an issue:
```
auto waybar::modules::Clock::update() -> void {
  const auto* tz = tzList_[tzCurrIdx_] != nullptr ? tzList_[tzCurrIdx_] : local_zone();
  const zoned_time now{tz, floor<seconds>(system_clock::now())};

  label_.set_markup(fmt_lib::vformat(m_locale_, format_, fmt_lib::make_format_args(now)));
...
```
Where I had the following values:
```
(gdb) p m_locale_
$1 = std::locale = "en_US.UTF-8"
(gdb) p format_
$2 = "{:%A  %b %e, %Y  %-I:%M %p}"
(gdb) p now
$3 = std::chrono::zoned_time = { "America/Los_Angeles" 1766180659s [2025-12-19 21:44:19] }
```

I couldn't reproduce with a minimal example that uses fmt (though not a perfect use case match because I had trouble with `zoned_time`, but I think that's neither here nor there because the actually `%-I` formatting didn't have a problem):
```
#include <fmt/chrono.h>

int main() {
    auto now = std::chrono::system_clock::now();
    auto formatted = fmt::vformat(std::locale("en_US.UTF-8"), std::string("{:%-I:%M %p}"), fmt::make_format_args(now));
    fmt::print("formatted at runtime by fmt: {}\n", formatted);
}
```
Compiled with
```
g++ hello-fmt.cpp -o hello-fmt -lfmt
```
Had no issue:
```
$ ./hello-fmt
formatted at runtime by fmt: 1:42 PM
```

However when running this similar example that uses c++23 (not using fmt):
```
#include <chrono>
#include <print>

int main() {
    auto now = std::chrono::system_clock::now();
    std::chrono::zoned_time zt { std::chrono::current_zone(), now };
    auto formatted = std::vformat(std::locale("en_US.UTF-8"), std::string("{:%-I:%M %p}"), std::make_format_args(now));
    std::print("formatted at runtime: {}\n", formatted);
}
```
When compiling with:
```
g++ hello-chrono-vformat-23.cpp -o hello-chrono-vformat-23 --std=c++23
```
I got the same runtime error:
```
$ ./hello-chrono-vformat-23
terminate called after throwing an instance of 'std::format_error'
  what():  chrono format error: invalid  specifier in chrono-specs
Aborted                    (core dumped) ./hello-chrono-vformat-23
```

So it would appear that my Waybar isn't using fmt at all, but C++23, and I guess C++20/C++23 time format specifiers don't have support for `-` to strip leading zeros.

In other words, this may be a building / packaging issue. I'm using Arch linux.

**Comment by carmanaught** [comment]
Following on a bit from your testing @jayache80, and I think essentially in agreement with your assessment, I've noticed the following.

Using version 0.15.0, if we trace things backwards a bit, you mention an issue around `fmt_lib::vformat` and near the top of [src/modules/clock.cpp](../blob/0.15.0/src/modules/clock.cpp) we have the following:
```cpp
namespace fmt_lib = waybar::util::date::format;
```

If we trace that back, we find that in [include/util/date.hpp](../blob/0.15.0/include/util/date.hpp) there is the following bit of code:
```cpp
// Format
namespace waybar::util::date::format {
#if HAVE_CHRONO_TIMEZONES
using namespace std;
#else
using namespace fmt;
#endif
}  // namespace waybar::util::date::format
```

The `HAVE_CHRONO_TIMEZONES` then seems to trace back to the [meson.build](../blob/0.15.0/meson.build), which has:
```meson
cpp_lib_chrono = compiler.compute_int('__cpp_lib_chrono', prefix : '#include <chrono>')
have_chrono_timezones = cpp_lib_chrono >= 201611
```

This, as I understand it, is what impacts the build path, so to speak, and as I understand it, results in the fact that Waybar doesn't use `fmt` if the C++ version is newer than the specified version.

If you set that to `have_chrono_timezones = false` however, the fallback build path uses [include/modules/simpleclock.hpp](../blob/0.15.0/include/modules/simpleclock.hpp) instead of [include/modules/clock.hpp](../blob/0.15.0/include/modules/clock.hpp) and that fails (for me at least on EndeavourOS) due to `fmt::localtime` being deprecated with more recent `fmt` versions as I understand it (found mention of `localtime` deprecation in issue #4226), along with the simple clock not being as featureful ("Provides only date time functionality. That's it."), according to the [Clock module WiKi page](https://github.com/Alexays/Waybar/wiki/Module:-Clock)).

In the end, I don't understand C++ nearly well enough to know what to do here to potentially get it to use `fmt` instead (if that's the correct path), but I do note that the [Clock module WiKi page](https://github.com/Alexays/Waybar/wiki/Module:-Clock) references format options and links to [https://fmt.dev/latest/syntax/#chrono-format-specifications](https://fmt.dev/latest/syntax/#chrono-format-specifications), which does mention about the padding modifiers and that seems to now be inconsistent with what Waybar actually supports.

**Comment by Alexays** [comment]
Closing — this was resolved/answered in the thread (environment, upstream, or a library limitation — not a Waybar bug still open). Feel free to reopen if you're still stuck. 🙏

**Comment by bisby** [comment]
I haven't used sway in a while, but I popped over just to see how this was doing. Still get the same issue as jayache, where including the `-` causes the clock to not load at all with the chrono spec error:

```
[error] clock: chrono format error: invalid  specifier in chrono-specs
```

This was on a fresh install of sway and waybar v0.15.0 on arch. I don't know what I have to do to make it use the right fmt library, or if that's baked in by arch at build time, but out of the box on a fresh install it's not working.

Doesn't really impact me either way anymore, but I don't feel like I really have clarity on what the actual issue is here still. 🤷🏻

**Comment by Alexays** [comment]
Thanks for checking back, @bisby. The `%-I`/`%OI` no-leading-zero modifiers genuinely aren't implemented by the underlying date/std::chrono formatter, so the spec is rejected — that part is an upstream limitation. That said, the module failing to load entirely on a bad spec (rather than warning and carrying on) is a separate, fair robustness concern; I'll look at making the clock degrade gracefully there. For now, `{:%I}` (with the leading zero) or a `{:%l}`-based workaround via a custom script is the way around it.
